app/src/main/kotlin/com/google/samples/apps/nowinandroid/MainActivity.kt

```diff
@@ -81,58 +74,64 @@ class MainActivity : ComponentActivity() {
     lateinit var userNewsResourceRepository: UserNewsResourceRepository
 
     val viewModel: MainActivityViewModel by viewModels()
-
     override fun onCreate(savedInstanceState: Bundle?) {
         val splashScreen = installSplashScreen()
         super.onCreate(savedInstanceState)
 
-        var uiState: MainActivityUiState by mutableStateOf(Loading)
+        // We keep this as a mutable state, so that we can track changes inside the composition.
+        // This allows us to react to dark/light mode changes.
+        var themeSettings by mutableStateOf(
```

Comment:
In practice aren't a bunch of these `staticCompositionLocalOf` which would cause a full recomposition anyway?

https://developer.android.com/reference/kotlin/androidx/compose/runtime/package-summary#staticCompositionLocalOf(kotlin.Function0)

Would this be simpler if you setKeepOnScreenCondition and maybe a conditional to wait for the themeSetting before composing any content?

Reply:
Sorry, I'm not sure I understand fully what you're asking 🤔 .
I think this is a bit different than `staticCompositionLocalOf`, because the nested composables can be skipped if params haven't changed. Though given we change the theme settings, maybe this would change the whole tree anyway (or most of it).

We're actually using the `setKeepOnScreenCondition` and I was trying to prevent composing anything if the state is not `Success`, but that didn't show better results.

This way we can improve the happy-path (default settings) to only compose once.

Reply:
Yeah, `staticCompositionLocalOf` is basically a global reset.  Your results and benchmarks will vary depending on whether the initial value

```
            ThemeSettings(
                darkTheme = resources.configuration.isSystemInDarkTheme,
                androidTheme = false,
                disableDynamicTheming = true,
            ),
```

is different in a way from the final

```
                    ThemeSettings(
                        darkTheme = uiState.shouldUseDarkTheme(systemDark),
                        androidTheme = uiState.shouldUseAndroidTheme,
                        disableDynamicTheming = uiState.shouldDisableDynamicTheming,
                    )
```

So it seems like the effect of the optimisation is dependent on these values.

So yeah, I was wondering if the correct solution is to wait until you have uiState before removing the splashscreen?

Reply:
You're right that it depends on the `ThemeSettings` state.
The thing is, we actually wait and keep the splash screen on https://github.com/android/nowinandroid/pull/1369/files#diff-705d6b4cbfd6b5724912a84fedddac4c41309ddfb4c5f0950e6e7bf1d6d4371aR132

Reply:
Can you make the default value null then?

And just skip drawing if null, which it shouldn't be.

Reply:
I tried preventing to compose anything if the `uiState != Success`, but then I realised there's a bunch of stuff that would be postponed and done the first time (like `rememberNiaAppState`, the composition locals and even stuff within the `NiaApp`).

If this recomposes second time, some composables can be skipped, so it should be shorter.
If we recompose everything the first time, we're just delaying it in time.